File: 99-use-cases/11-create-eval-flag/go/main.go
// Console grid navigator — create and evaluate a single LaunchDarkly highlight flag.
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
	ld "github.com/launchdarkly/go-server-sdk/v7"
	"golang.org/x/term"
)

const (
	appBanner        = "11-create-eval-flag[go]"
	flagHighlight    = "configure-grid-selection-green-highlight"
	defaultHighlight = "none"
)

var (
	rows        = [3]string{"t", "m", "b"}
	cols        = [3]string{"l", "m", "r"}
	validColors = map[string]struct{}{
		"yellow": {}, "red": {}, "blue": {}, "green": {}, "purple": {},
	}
	ldClient *ld.LDClient
)

const (
	bgReset = "\033[48;5;236m"
	reset   = "\033[0m"
)

type flagValues struct {
	Username       string `json:"username"`
	FlagValue      string `json:"flagValue"`
	HighlightColor string `json:"highlightColor"`
	ColorLabel     string `json:"colorLabel"`
}

type position struct {
	row, col int
}

type moveResult struct {
	row, col int
	moved    bool
}

type sessionAction int

const (
	actionQuit sessionAction = iota
	actionLogout
)

func initLaunchDarkly() {
	sdkKey := os.Getenv("LD_SDK_KEY")
	if sdkKey == "" {
		fmt.Fprintln(os.Stderr, "Warning: LD_SDK_KEY not set — highlight defaults to none.")
		return
	}
	client, err := ld.MakeClient(sdkKey, 10*time.Second)
	if err != nil {
		fmt.Fprintln(os.Stderr, "Warning: LaunchDarkly SDK did not initialize — highlight defaults to none.")
		return
	}
	ldClient = client
}

func normalizeHighlightColor(raw interface{}) string {
	color := strings.ToLower(strings.TrimSpace(fmt.Sprint(raw)))
	if color == "" {
		color = "none"
	}
	if _, ok := validColors[color]; ok {
		return color
	}
	return "none"
}

func colorLabel(highlightColor string) string {
	if highlightColor == "none" {
		return "(no-color)"
	}
	return fmt.Sprintf("(%s)", highlightColor)
}

func buildResponse(username string, raw interface{}) flagValues {
	color := normalizeHighlightColor(raw)
	flagValue := strings.TrimSpace(fmt.Sprint(raw))
	if flagValue == "" {
		flagValue = "none"
	}
	return flagValues{
		Username:       username,
		FlagValue:      flagValue,
		HighlightColor: color,
		ColorLabel:     colorLabel(color),
	}
}

func evaluateHighlight(username string) flagValues {
	if ldClient == nil || username == "" {
		return buildResponse(username, defaultHighlight)
	}
	context := ldcontext.NewBuilder(username).Build()
	raw, _ := ldClient.StringVariation(flagHighlight, context, defaultHighlight)
	return buildResponse(username, raw)
}

func ansiColor(color string) string {
	switch color {
	case "yellow":
		return "\033[93m"
	case "red":
		return "\033[91m"
	case "blue":
		return "\033[94m"
	case "green":
		return "\033[92m"
	case "purple":
		return "\033[35m"
	default:
		return ""
	}
}

func colorize(text, color string) string {
	if color == "" || color == "none" {
		return text
	}
	ansi := ansiColor(color)
	if ansi == "" {
		return text
	}
	return ansi + text + reset + bgReset
}

func formatPos(row, col int) string {
	return rows[row] + "/" + cols[col]
}

func tryMove(row, col, dr, dc int) moveResult {
	nr := clamp(row+dr, 0, 2)
	nc := clamp(col+dc, 0, 2)
	if nr == row && nc == col {
		return moveResult{row, col, false}
	}
	return moveResult{nr, nc, true}
}

func clamp(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

func readUsername(reader *bufio.Reader) (string, error) {
	fmt.Println(appBanner)
	fmt.Println("Login\n")
	for {
		fmt.Print("Username: ")
		line, err := reader.ReadString('\n')
		if err != nil {
			return "", err
		}
		name := strings.TrimSpace(line)
		if name != "" {
			return name, nil
		}
		fmt.Println("Username is required.")
	}
}

func drawCell(selected bool, line int, color string) string {
	if !selected {
		switch line {
		case 0:
			return "┌───┐"
		case 1:
			return "│   │"
		default:
			return "└───┘"
		}
	}
	var text string
	switch line {
	case 0:
		text = "┏━━━┓"
	case 1:
		text = "┃ X ┃"
	default:
		text = "┗━━━┛"
	}
	if color != "" && color != "none" {
		return colorize(text, color)
	}
	return text
}

func writeLine(out *strings.Builder, line string) {
	out.WriteString(line)
	out.WriteString("\r\n")
}

func render(username string, row, col int, previous *position, flags flagValues) {
	var out strings.Builder
	out.WriteString("\033[H\033[2J")
	out.WriteString(bgReset)

	prevText := "—"
	if previous != nil {
		prevText = formatPos(previous.row, previous.col)
	}

	writeLine(&out, appBanner)
	nameLine := fmt.Sprintf(
		"Name: %s%s",
		colorize(username, flags.HighlightColor),
		colorize(" "+flags.ColorLabel, flags.HighlightColor),
	)
	writeLine(&out, nameLine+reset+bgReset)
	writeLine(&out, fmt.Sprintf("Flag value: %s", flags.FlagValue))
	writeLine(&out, fmt.Sprintf("Current position: %s", formatPos(row, col)))
	writeLine(&out, fmt.Sprintf("Previous position: %s", prevText))
	writeLine(&out, "")
	writeLine(&out, "Use arrow keys or WASD to move (L to logout, Q to quit).")
	writeLine(&out, "Toggle the flag in LaunchDarkly — changes appear within ~1s.")
	writeLine(&out, "")

	cellColor := flags.HighlightColor
	for r := 0; r < 3; r++ {
		top := make([]string, 3)
		mid := make([]string, 3)
		bot := make([]string, 3)
		for c := 0; c < 3; c++ {
			selected := r == row && c == col
			highlight := "none"
			if selected && cellColor != "none" {
				highlight = cellColor
			}
			top[c] = drawCell(selected, 0, highlight)
			mid[c] = drawCell(selected, 1, highlight)
			bot[c] = drawCell(selected, 2, highlight)
		}
		writeLine(&out, strings.Join(top, " "))
		writeLine(&out, strings.Join(mid, " "))
		writeLine(&out, strings.Join(bot, " "))
	}

	fmt.Print(out.String())
}

func readKey(reader *bufio.Reader) (dr, dc int, action sessionAction, endSession, ok bool) {
	b, err := reader.ReadByte()
	if err != nil {
		return 0, 0, actionQuit, true, false
	}
	switch b {
	case 3, 'q', 'Q':
		return 0, 0, actionQuit, true, true
	case 'l', 'L':
		return 0, 0, actionLogout, true, true
	case 'w', 'W':
		return -1, 0, 0, false, true
	case 's', 'S':
		return 1, 0, 0, false, true
	case 'a', 'A':
		return 0, -1, 0, false, true
	case 'd', 'D':
		return 0, 1, 0, false, true
	case 27:
		b2, err := reader.ReadByte()
		if err != nil || b2 != '[' {
			return 0, 0, 0, false, false
		}
		b3, err := reader.ReadByte()
		if err != nil {
			return 0, 0, 0, false, false
		}
		switch b3 {
		case 'A':
			return -1, 0, 0, false, true
		case 'B':
			return 1, 0, 0, false, true
		case 'C':
			return 0, 1, 0, false, true
		case 'D':
			return 0, -1, 0, false, true
		}
	}
	return 0, 0, 0, false, false
}

func runGrid(username string, reader *bufio.Reader) sessionAction {
	row, col := 1, 1
	var previous *position
	lastEval := time.Time{}

	for {
		flags := evaluateHighlight(username)
		render(username, row, col, previous, flags)
		lastEval = time.Now()

		for time.Since(lastEval) < 500*time.Millisecond {
			if reader.Buffered() > 0 {
				break
			}
			time.Sleep(50 * time.Millisecond)
		}
		if reader.Buffered() == 0 && time.Since(lastEval) >= 500*time.Millisecond {
			continue
		}

		dr, dc, action, endSession, ok := readKey(reader)
		if !ok {
			continue
		}
		if endSession {
			return action
		}
		result := tryMove(row, col, dr, dc)
		if result.moved {
			prev := position{row, col}
			previous = &prev
			row, col = result.row, result.col
		}
	}
}

func main() {
	if len(os.Args) >= 3 && os.Args[1] == "--evaluate-once" {
		initLaunchDarkly()
		defer func() {
			if ldClient != nil {
				ldClient.Close()
			}
		}()
		result := evaluateHighlight(os.Args[2])
		enc := json.NewEncoder(os.Stdout)
		enc.SetEscapeHTML(false)
		_ = enc.Encode(result)
		return
	}

	initLaunchDarkly()
	defer func() {
		if ldClient != nil {
			ldClient.Close()
		}
	}()

	reader := bufio.NewReader(os.Stdin)
	fd := int(os.Stdin.Fd())

	for {
		username, err := readUsername(reader)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		oldState, err := term.MakeRaw(fd)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}

		action := runGrid(username, reader)
		term.Restore(fd, oldState)

		if action == actionQuit {
			break
		}
	}
}

File: 99-use-cases/11-create-eval-flag/rust/src/main.rs
//! Console grid navigator — create and evaluate a single LaunchDarkly highlight flag.

use crossterm::{
    cursor::{Hide, MoveTo, Show},
    event::{self, Event, KeyCode, KeyEvent, KeyModifiers},
    execute, queue,
    style::{Color, Print, SetBackgroundColor, SetForegroundColor},
    terminal::{self, ClearType},
};
use launchdarkly_server_sdk::{Client, ConfigBuilder, ContextBuilder};
use std::collections::HashSet;
use std::io::{self, Write};
use std::sync::Arc;
use std::time::{Duration, Instant};

const FLAG_HIGHLIGHT: &str = "configure-grid-selection-green-highlight";
const DEFAULT_HIGHLIGHT: &str = "none";

const ROWS: [&str; 3] = ["t", "m", "b"];
const APP_BANNER: &str = "11-create-eval-flag[rust]";

const COLS: [&str; 3] = ["l", "m", "r"];

struct FlagValues {
    username: String,
    flag_value: String,
    highlight_color: String,
    color_label: String,
}

struct Position {
    row: i32,
    col: i32,
}

struct MoveResult {
    row: i32,
    col: i32,
    moved: bool,
}

#[derive(PartialEq, Eq)]
enum SessionAction {
    Quit,
    Logout,
}

struct App {
    client: Option<Arc<Client>>,
}

impl App {
    fn new() -> Self {
        let client = std::env::var("LD_SDK_KEY").ok().and_then(|key| {
            let config = ConfigBuilder::new(&key).build().ok()?;
            let client = Client::build(config).ok()?;
            client.start_with_runtime().ok()?;

            let deadline = Instant::now() + Duration::from_secs(10);
            while !client.initialized() && Instant::now() < deadline {
                std::thread::sleep(Duration::from_millis(50));
            }
            if client.initialized() {
                Some(Arc::new(client))
            } else {
                eprintln!("Warning: LaunchDarkly SDK did not initialize — highlight defaults to none.");
                None
            }
        });
        if std::env::var("LD_SDK_KEY").is_err() {
            eprintln!("Warning: LD_SDK_KEY not set — highlight defaults to none.");
        }
        Self { client }
    }

    fn evaluate_highlight(&self, username: &str) -> FlagValues {
        let Some(client) = &self.client else {
            return build_response(username, DEFAULT_HIGHLIGHT);
        };
        let context = ContextBuilder::new(username).build().unwrap_or_else(|_| {
            ContextBuilder::new("anonymous").build().unwrap()
        });
        let raw = client.str_variation(
            &context,
            FLAG_HIGHLIGHT,
            DEFAULT_HIGHLIGHT.to_string(),
        );
        build_response(username, &raw)
    }
}

impl Drop for App {
    fn drop(&mut self) {
        if let Some(client) = self.client.take() {
            if let Ok(client) = Arc::try_unwrap(client) {
                client.close();
            }
        }
    }
}

fn color_names() -> HashSet<&'static str> {
    ["yellow", "red", "blue", "green", "purple"]
        .into_iter()
        .collect()
}

fn normalize_highlight_color(raw: &str) -> String {
    let color = raw.trim().to_lowercase();
    if color_names().contains(color.as_str()) {
        color
    } else {
        "none".to_string()
    }
}

fn color_label(highlight_color: &str) -> String {
    if highlight_color == "none" {
        "(no-color)".to_string()
    } else {
        format!("({highlight_color})")
    }
}

fn build_response(username: &str, raw: &str) -> FlagValues {
    let color = normalize_highlight_color(raw);
    let flag_value = if raw.trim().is_empty() {
        "none".to_string()
    } else {
        raw.trim().to_string()
    };
    FlagValues {
        username: username.to_string(),
        flag_value,
        highlight_color: color.clone(),
        color_label: color_label(&color),
    }
}

fn term_color(color: &str) -> Option<Color> {
    match color {
        "yellow" => Some(Color::Yellow),
        "red" => Some(Color::Red),
        "blue" => Some(Color::Blue),
        "green" => Some(Color::Green),
        "purple" => Some(Color::Magenta),
        _ => None,
    }
}

fn format_pos(row: i32, col: i32) -> String {
    format!("{}/{}", ROWS[row as usize], COLS[col as usize])
}

fn try_move(row: i32, col: i32, dr: i32, dc: i32) -> MoveResult {
    let nr = (row + dr).clamp(0, 2);
    let nc = (col + dc).clamp(0, 2);
    if nr == row && nc == col {
        MoveResult { row, col, moved: false }
    } else {
        MoveResult {
            row: nr,
            col: nc,
            moved: true,
        }
    }
}

fn read_username() -> io::Result<String> {
    println!("{APP_BANNER}");
    println!("Login\n");
    loop {
        print!("Username: ");
        io::stdout().flush()?;
        let mut line = String::new();
        io::stdin().read_line(&mut line)?;
        let name = line.trim().to_string();
        if !name.is_empty() {
            return Ok(name);
        }
        println!("Username is required.");
    }
}

fn cell_line(selected: bool, line: usize) -> &'static str {
    if selected {
        match line {
            0 => "┏━━━┓",
            1 => "┃ X ┃",
            _ => "┗━━━┛",
        }
    } else {
        match line {
            0 => "┌───┐",
            1 => "│   │",
            _ => "└───┘",
        }
    }
}

fn print_colored_line(
    out: &mut impl Write,
    y: u16,
    text: &str,
    color: Option<Color>,
) -> io::Result<()> {
    queue!(out, MoveTo(0, y), SetBackgroundColor(Color::Rgb { r: 30, g: 30, b: 46 }))?;
    if let Some(c) = color {
        queue!(out, SetForegroundColor(c), Print(text), SetForegroundColor(Color::Reset))?;
    } else {
        queue!(out, Print(text))?;
    }
    Ok(())
}

fn render(
    out: &mut impl Write,
    username: &str,
    row: i32,
    col: i32,
    previous: Option<&Position>,
    flags: &FlagValues,
) -> io::Result<()> {
    execute!(out, MoveTo(0, 0), terminal::Clear(ClearType::All))?;

    let prev_text = previous
        .map(|p| format_pos(p.row, p.col))
        .unwrap_or_else(|| "—".to_string());

    let highlight = term_color(&flags.highlight_color);
    let name_line = format!("Name: {} {}", username, flags.color_label);

    let mut y = 0u16;
    print_colored_line(out, y, APP_BANNER, None)?;
    y += 1;
    print_colored_line(out, y, &name_line, highlight)?;
    y += 1;
    print_colored_line(out, y, &format!("Flag value: {}", flags.flag_value), None)?;
    y += 1;
    print_colored_line(out, y, &format!("Current position: {}", format_pos(row, col)), None)?;
    y += 1;
    print_colored_line(out, y, &format!("Previous position: {prev_text}"), None)?;
    y += 2;
    print_colored_line(out, y, "Use arrow keys or WASD to move (L to logout, Q to quit).", None)?;
    y += 1;
    print_colored_line(out, y, "Toggle the flag in LaunchDarkly — changes appear within ~1s.", None)?;
    y += 2;

    let cell_color = if flags.highlight_color == "none" {
        None
    } else {
        term_color(&flags.highlight_color)
    };

    for r in 0..3 {
        for line_idx in 0..3 {
            let parts: Vec<String> = (0..3)
                .map(|c| {
                    let selected = r == row && c == col;
                    let text = cell_line(selected, line_idx);
                    if selected && cell_color.is_some() {
                        format!("\x1b[{}m{text}\x1b[0m", match cell_color {
                            Some(Color::Yellow) => "93",
                            Some(Color::Red) => "91",
                            Some(Color::Blue) => "94",
                            Some(Color::Green) => "92",
                            Some(Color::Magenta) => "35",
                            _ => "0",
                        })
                    } else {
                        text.to_string()
                    }
                })
                .collect();
            print_colored_line(out, y, &parts.join(" "), None)?;
            y += 1;
        }
    }

    out.flush()?;
    Ok(())
}

fn run_grid(out: &mut impl Write, app: &App, username: &str) -> io::Result<SessionAction> {
    let mut row = 1;
    let mut col = 1;
    let mut previous: Option<Position> = None;
    let mut last_eval;

    loop {
        let flags = app.evaluate_highlight(username);
        render(out, username, row, col, previous.as_ref(), &flags)?;
        last_eval = Instant::now();

        loop {
            if last_eval.elapsed() >= Duration::from_millis(500) {
                break;
            }
            if event::poll(Duration::from_millis(50))? {
                break;
            }
        }

        if last_eval.elapsed() >= Duration::from_millis(500) && !event::poll(Duration::from_millis(0))? {
            continue;
        }

        if let Event::Key(KeyEvent { code, modifiers, .. }) = event::read()? {
            if modifiers.contains(KeyModifiers::CONTROL) {
                return Ok(SessionAction::Quit);
            }
            let movement = match code {
                KeyCode::Char('q') | KeyCode::Char('Q') => return Ok(SessionAction::Quit),
                KeyCode::Char('l') | KeyCode::Char('L') => return Ok(SessionAction::Logout),
                KeyCode::Up | KeyCode::Char('w') | KeyCode::Char('W') => Some((-1, 0)),
                KeyCode::Down | KeyCode::Char('s') | KeyCode::Char('S') => Some((1, 0)),
                KeyCode::Left | KeyCode::Char('a') | KeyCode::Char('A') => Some((0, -1)),
                KeyCode::Right | KeyCode::Char('d') | KeyCode::Char('D') => Some((0, 1)),
                _ => None,
            };
            let Some((dr, dc)) = movement else {
                continue;
            };
            let result = try_move(row, col, dr, dc);
            if result.moved {
                previous = Some(Position { row, col });
                row = result.row;
                col = result.col;
            }
        }
    }
}

fn main() -> io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() >= 3 && args[1] == "--evaluate-once" {
        let app = App::new();
        let result = app.evaluate_highlight(&args[2]);
        println!(
            "{{\"username\":\"{}\",\"flagValue\":\"{}\",\"highlightColor\":\"{}\",\"colorLabel\":\"{}\"}}",
            result.username.replace('\\', "\\\\").replace('"', "\\\""),
            result.flag_value.replace('\\', "\\\\").replace('"', "\\\""),
            result.highlight_color.replace('\\', "\\\\").replace('"', "\\\""),
            result.color_label.replace('\\', "\\\\").replace('"', "\\\""),
        );
        return Ok(());
    }

    let app = App::new();
    let mut stdout = io::stdout();

    loop {
        let username = read_username()?;
        terminal::enable_raw_mode()?;
        execute!(stdout, Hide)?;

        let action = run_grid(&mut stdout, &app, &username)?;

        execute!(stdout, Show)?;
        terminal::disable_raw_mode()?;

        if action == SessionAction::Quit {
            break;
        }
    }

    Ok(())
}
